tests/integration/ha_test.go

```diff
@@ -256,6 +270,32 @@ func waitLines(t *testing.T, q Querier, num int, timeout time.Duration) error {
 	return fmt.Errorf("timeout waiting for %d lines, got: %d", num, c)
 }
 
+// waitForLine waits until a row with the expected ID appears in table01
+func waitForLine(t *testing.T, q Querier, expected int, timeout time.Duration) error {
```

Comment:
This was what I expected `waitForLines` to do! Tripped myself up by assuming its behaviour. Two small things here:

- I've tried to comment my new test helpers to make it clear why you might use them, and how they behave. I would have found it really useful to have comments like these on the ones that already exist when getting familiar with the test suite, though I totally understand when this is mostly a one man operation why they wouldn't be there
- I'm a bit worried `waitLines` might be too loose an assertion in some tests. It could be that we're passing tests when a specific token value never replicated, while a subsequent one did. Not sure on this but I'd normally avoid this type of assertion in favour of something more specific, like this `waitForLine`

Reply:
I think the result is the same just with different logic:

What the tests really want to check is that all the transactions are applied. Waiting for the number of rows in an append only table is a way to do this.

If you only append rows to a table you know what the expected total number should be. So I think waiting for the total number of rows is enough. Waiting for the last row you inserted is another way for doing the same thing. Or am I missing something?

I'll stick to just one of them to not add additional confusion the the integration tests (if you really prefer waiting for the last row feel free to create a PR that replaces it)




P.S. The real issue is the function name that will require a good s/line/row/ :smile:

Reply:
In this test, we're expecting to lose data. So we do three writes of 1, 2 and 3, but the final database should only ever get 1 and 3. 

If I use the count of lines then I'd have to expect there to be 2 lines, but that will pass even if I never received the correct write and only have 1 and 2 in the final standby (while the primary node has 1 and 3). 

Does that make any sense?